tq/ssh.go

```diff
@@ -0,0 +1,390 @@
+		msg := "no message provided"
+		if len(lines) > 0 {
+			msg = lines[0]
+		}
+		return nil, fmt.Errorf("batch response: status %d from server (%s)", status, msg)
+	}
+
+	sort.Strings(lines)
+	for _, line := range lines {
+		entries := strings.Split(line, " ")
+		if len(entries) < 3 {
+			return nil, fmt.Errorf("batch response: malformed response: %q", line)
+		}
+		length := len(bRes.Objects)
+		if length == 0 || bRes.Objects[length-1].Oid != entries[0] {
+			bRes.Objects = append(bRes.Objects, &Transfer{Actions: make(map[string]*Action)})
+		}
+		transfer := bRes.Objects[len(bRes.Objects)-1]
+		transfer.Oid = entries[0]
+		transfer.Size, err = strconv.ParseInt(entries[1], 10, 64)
+		if err != nil {
+			return nil, fmt.Errorf("batch response: invalid size: %s", entries[1])
+		}
+		if entries[2] == "noop" {
+			continue
+		}
+		transfer.Actions[entries[2]] = &Action{}
+		if len(entries) > 3 {
+			for _, entry := range entries[3:] {
+				if strings.HasPrefix(entry, "id=") {
+					transfer.Actions[entries[2]].Id = entry[3:]
+				} else if strings.HasPrefix(entry, "token=") {
+					transfer.Actions[entries[2]].Token = entry[6:]
+				} else if strings.HasPrefix(entry, "expires-in=") {
+					transfer.Actions[entries[2]].ExpiresIn, err = strconv.Atoi(entry[11:])
+					if err != nil {
+						return nil, fmt.Errorf("batch response: invalid expires-in: %s", entry)
+					}
+				} else if strings.HasPrefix(entry, "expires-at=") {
+					transfer.Actions[entries[2]].ExpiresAt, err = time.Parse(time.RFC3339, entry[11:])
+					if err != nil {
+						return nil, fmt.Errorf("batch response: invalid expires-at: %s", entry)
+					}
+				}
+			}
+		}
+	}
+
+	for _, obj := range bRes.Objects {
+		obj.Missing = missing[obj.Oid]
+		for _, a := range obj.Actions {
+			a.createdAt = requestedAt
+		}
+	}
+
+	return bRes, nil
+}
+
+func (a *SSHBatchTransferAdapter) MaxRetries() int {
+	return a.maxRetries
+}
+
+func (a *SSHBatchTransferAdapter) SetMaxRetries(n int) {
+	a.maxRetries = n
+}
+
+type SSHAdapter struct {
+	*adapterBase
+	ctx      lfshttp.Context
+	transfer *ssh.SSHTransfer
+}
+
+// WorkerStarting is called when a worker goroutine starts to process jobs
+// Implementations can run some startup logic here & return some context if needed
+func (a *SSHAdapter) WorkerStarting(workerNum int) (interface{}, error) {
+	a.transfer.SetConnectionCountAtLeast(workerNum + 1)
+	return a.transfer.Connection(workerNum), nil
+}
+
+// WorkerEnding is called when a worker goroutine is shutting down
+// Implementations can clean up per-worker resources here, context is as returned from WorkerStarting
+func (a *SSHAdapter) WorkerEnding(workerNum int, ctx interface{}) {
+}
+
+func (a *SSHAdapter) tempDir() string {
+	// Must be dedicated to this adapter as deleted by ClearTempStorage
+	// Also make local to this repo not global, and separate to localstorage temp,
+	// which gets cleared at the end of every invocation
+	d := filepath.Join(a.fs.LFSStorageDir, "incomplete")
```

Comment:
No, there isn't any reason not to share them.  The directories can be safely shared, and it's not even like both could legitimately be used in the same invocation anyway, since one will never use both adapters at the same time.  I've updated the comments.